File: backend/app/model/user_meta.py
from sqlalchemy import func, Column, Integer, ForeignKey, DateTime, Boolean

from app.database import Base
from app.model.participant_relationship import Relationship


class UserMeta(Base):
    __tablename__ = "usermeta"
    __label__ = "User Meta Info"
    id = Column(Integer, ForeignKey("stardrive_user.id"), primary_key=True)
    last_updated = Column(DateTime(timezone=True), default=func.now())
    self_participant = Column(Boolean)
    self_has_guardian = Column(Boolean)
    guardian = Column(Boolean)
    guardian_has_dependent = Column(Boolean)
    professional = Column(Boolean)
    interested = Column(Boolean)

    def get_relationship(self):
        if self.self_participant:
            return None if self.self_has_guardian else Relationship.self_participant.name
        if self.guardian and self.guardian_has_dependent:
            return Relationship.self_guardian.name
        if self.professional:
            return Relationship.self_professional.name
        # Lower Precedence Relationships
        if self.guardian and not self.guardian_has_dependent:
            return Relationship.self_interested.name
        if self.interested:
            return Relationship.self_interested.name
        return ""

File: frontend/src/app/shared/models/user_meta.ts
import {ParticipantRelationship} from './participantRelationship';

export class UserMeta {
  id: number;
  last_updated?: Date;
  self_participant?: boolean;
  self_has_guardian?: boolean;
  guardian?: boolean;
  guardian_has_dependent?: boolean;
  professional?: boolean;
  interested?: boolean;
  self_relationship: ParticipantRelationship;

  constructor(_props: {}) {
    for (const propName in _props) {
      if (_props.hasOwnProperty(propName)) {
        this[propName] = _props[propName];
      }
    }
  }
}
